README.md
# Gitlab exporter

Simple prometheus exporter that exposes basic stats about
projects in gitlab.

## Installation

To install gitlab exporter simply grab a binary from the [releases page](https://github.com/j6s/prometheus-gitlab-exporter/releases).

## Stats

The following is an example of the stats collected by this exporter.
`gitlab_last_update` contains the timestamp of the last time the data
has been updated from the gitlab API: Data is not updated everytime
the `/metrics` endpoint is called - it is updated in regular intervals
(every 5 minutes by default) in order to not produce to many API
requests.

```
gitlab_last_update 1529090327

gitlab_project_stars{repo="namespace1___project1"} 0
gitlab_project_forks{repo="namespace1___project1"} 0
gitlab_project_commit_count{repo="namespace1___project1"} 0
gitlab_project_storage_size{repo="namespace1___project1"} 0
gitlab_project_repository_size{repo="namespace1___project1"} 0
gitlab_project_lfs_object_size{repo="namespace1___project1"} 0
gitlab_project_job_artifacts_size{repo="namespace1___project1"} 0
gitlab_project_last_activity{repo="namespace1___project1"} 1529046202

gitlab_project_stars{repo="namespace2___project2"} 0
gitlab_project_forks{repo="namespace2___project2"} 0
gitlab_project_commit_count{repo="namespace2___project2"} 11
gitlab_project_storage_size{repo="namespace2___project2"} 94371
gitlab_project_repository_size{repo="namespace2___project2"} 94371
gitlab_project_lfs_object_size{repo="namespace2___project2"} 0
gitlab_project_job_artifacts_size{repo="namespace2___project2"} 0
gitlab_project_last_activity{repo="namespace2___project2"} 1528462352
```

## Usage

A personal access token is required in order to communicate with the API.
You can learn more about personal access tokens and how to manage them [in gitlab API documentation](https://docs.gitlab.com/ee/api/#personal-access-tokens).

```
$ prometheus-gitlab-exporter --help
Prometheus Gitlab exporter
============================
Simple exporter that exposes gitlab project statistics to prometheus.
https://github.com/j6s/prometheus-gitlab-exporter

Usage
-----
$ prometheus-gitlab-exporter --url='https://git.acme.org' --token='abcdef123'
$ prometheus-gitlab-exporter --url='https://git.acme.org' --token='abcdef123' --poll-interval='15m' --bind='hostname.com:9898'

Arguments
---------
  -bind string
    	Address to bind to. The service will be available at this address.
    	 (default ":8123")
  -poll-interval duration
    	Poll interval in minutes. The data will be updated every time interval
    	in order to avoid excessive API use.
    	Every string accepted by the golang time package is valid.
    	https://golang.org/pkg/time/#example_Duration
    	 (default 5m0s)
  -token string
    	The personal access token that is used to communicate with
    	the API. For more information about the Gitlab API please
    	refer to the GitLab documentation.
    	https://docs.gitlab.com/ee/api/#personal-access-tokens
  -url string
    	The base URL of the gitlab instance including protocol.
    	This string must not contain any path information other than the
    	index route of gitlab. If your server runs on a non-standard port
    	(not 80 or 443 for http and https) then you may specify it using
    	a colon.
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.0.1] 2018-06-15
### Added
- Now supplying binaries for major operating systems on the [releases page](https://github.com/j6s/prometheus-gitlab-exporter/releases)
### Fixed
- Fixed invalid label syntax that was not delimited by quotes.

## [1.0.0] 2018-06-14
### Added
- Initial release of the tool. Literally everthing was added.


project.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"strconv"
	"time"
	"regexp"
)

type ProjectStats struct {
	CommitCount      int `json:"commit_count"`
	StorageSize      int `json:"storage_size"`
	RepositorySize   int `json:"repository_size"`
	LfsObjectSize    int `json:"lfs_object_size"`
	JobArtifactsSize int `json:"job_artifacts_size"`
}

type Project struct {
	PathWithNamespace string       `json:"path_with_namespace"`
	StarCount         int          `json:"star_count"`
	ForkCount         int          `json:"fork_count"`
	OpenIssueCount    int          `json:"openIssueCount"`
	LastActivityAt    time.Time    `json:"last_activity_at"`
	Statistics        ProjectStats `json:"statistics"`
}

/**
 *	Extracts the stats from a single project into a
 *	prometheus compatible string.
 *
 *	@param project The project to extract stats from
 *	@return A prometheus style statistics string for the project
 */
func (project Project) PrometheusStats() string {
	path := regexp.MustCompile(`\/`).ReplaceAllString(project.PathWithNamespace, "___")
	stats := ""
	stats = fmt.Sprintf("%s\ngitlab_project_stars{repo=\"%s\"} %d", stats, path, project.StarCount)
	stats = fmt.Sprintf("%s\ngitlab_project_forks{repo=\"%s\"} %d", stats, path, project.ForkCount)
	stats = fmt.Sprintf("%s\ngitlab_project_commit_count{repo=\"%s\"} %d", stats, path, project.Statistics.CommitCount)
	stats = fmt.Sprintf("%s\ngitlab_project_storage_size{repo=\"%s\"} %d", stats, path, project.Statistics.StorageSize)
	stats = fmt.Sprintf("%s\ngitlab_project_repository_size{repo=\"%s\"} %d", stats, path, project.Statistics.RepositorySize)
	stats = fmt.Sprintf("%s\ngitlab_project_lfs_object_size{repo=\"%s\"} %d", stats, path, project.Statistics.LfsObjectSize)
	stats = fmt.Sprintf("%s\ngitlab_project_job_artifacts_size{repo=\"%s\"} %d", stats, path, project.Statistics.JobArtifactsSize)
	stats = fmt.Sprintf("%s\ngitlab_project_last_activity{repo=\"%s\"} %d", stats, path, project.LastActivityAt.Unix())
	return stats
}

/**
 *	Fetches all projects from the configured gitlab endpoint.
 *
 *	@return A list of all projects known to the current gitlab instance
 */
func GetRepositories(gitlabUrl string, token string) []Project {
	projects := make([]Project, 0)
	page := 1

	for true {
		// Fetch a page from the API.
		projectsUrl := fmt.Sprintf("%s/api/v4/projects?private_token=%s&per_page=100&statistics=1&page=%d", gitlabUrl, token, page)
		log.Printf("Requesting %s\n", projectsUrl)
		response, error := http.Get(projectsUrl)
		if error != nil {
			log.Fatalf(error.Error())
		}

		// Merge the results back to the complete array.
		projectsInPage := make([]Project, 0)
		error = json.NewDecoder(response.Body).Decode(&projectsInPage)
		if error != nil {
			log.Fatalf(error.Error())
		}
		projects = append(projects, projectsInPage...)

		// Parse the X-Next-Page Header in order to figure out
		// if another page should be requested.
		// If not, then return the current projects.
		pageHeader := response.Header["X-Next-Page"][0]
		if pageHeader == "" {
			return projects
		}
		page, error = strconv.Atoi(pageHeader)
		if error != nil {
			log.Fatalf(error.Error())
		}
	}

	return projects
}


main.go
package main

import (
	"flag"
	"fmt"
	"log"
	"net/http"
	"os"
	"time"
)

var (
	token = flag.String(
		"token",
		"",
		"The personal access token that is used to communicate with \n"+
			"the API. For more information about the Gitlab API please \n"+
			"refer to the GitLab documentation.\n"+
			"https://docs.gitlab.com/ee/api/#personal-access-tokens",
	)
	gitlabUrl = flag.String(
		"url",
		"",
		"The base URL of the gitlab instance including protocol.\n"+
			"This string must not contain any path information other than the\n"+
			"index route of gitlab. If your server runs on a non-standard port\n"+
			"(not 80 or 443 for http and https) then you may specify it using\n"+
			"a colon.",
	)
	pollInterval = flag.Duration(
		"poll-interval",
		5*time.Minute,
		"Poll interval in minutes. The data will be updated every time interval\n"+
			"in order to avoid excessive API use.\n"+
			"Every string accepted by the golang time package is valid.\n"+
			"https://golang.org/pkg/time/#example_Duration\n",
	)
	bind = flag.String(
		"bind",
		":8123",
		"Address to bind to. The service will be available at this address.\n",
	)
)

func main() {
	flag.Usage = func() {
		fmt.Fprintf(os.Stderr, "Prometheus Gitlab exporter \n")
		fmt.Fprintf(os.Stderr, "============================\n")
		fmt.Fprintf(os.Stderr, "Simple exporter that exposes gitlab project statistics to prometheus.\n")
		fmt.Fprintf(os.Stderr, "https://github.com/j6s/prometheus-gitlab-exporter\n")
		fmt.Fprintf(os.Stderr, "\nUsage\n")
		fmt.Fprintf(os.Stderr, "-----\n")
		fmt.Fprintf(os.Stderr, "$ prometheus-gitlab-exporter --url='https://git.acme.org' --token='abcdef123'\n")
		fmt.Fprintf(os.Stderr, "$ prometheus-gitlab-exporter --url='https://git.acme.org' --token='abcdef123' --poll-interval='15m' --bind='hostname.com:9898'\n")
		fmt.Fprintf(os.Stderr, "\nArguments\n")
		fmt.Fprintf(os.Stderr, "---------\n")
		flag.PrintDefaults()
	}
	flag.Parse()

	// Validate arguments
	if *gitlabUrl == "" && *token == "" {
		log.Fatalf("The arguments --token and --url must be set")
	}

	// Update the stats regularly
	stats := ""
	go func() {
		stats = getStats()
		log.Printf("Updating data every %v seconds", pollInterval.Seconds())
		for range time.Tick(*pollInterval) {
			stats = getStats()
		}
	}()

	// Start http server that responds with the stats
	router := http.NewServeMux()
	router.HandleFunc("/metrics", func(w http.ResponseWriter, req *http.Request) {
		fmt.Fprintf(w, "%s", stats)
	})

	log.Printf("Listening on port :8123")
	if err := http.ListenAndServe(*bind, router); err != nil {
		log.Printf("Could not bind webserver to %q: %v", *bind, err)
	}
}

func getStats() string {
	log.Printf("Updating")
	projects := GetRepositories(*gitlabUrl, *token)
	stats := fmt.Sprintf("gitlab_last_update %d\n", time.Now().Unix())

	for _, project := range projects {
		stats = fmt.Sprintf("%s\n%s\n", stats, project.PrometheusStats())
	}
	return stats
}
